src/v/security/audit/schemas/types.h

```diff
@@ -0,0 +1,481 @@
+    email_file_activity = 4011,
+    email_url_activity = 4012,
+    device_inventory_info = 5001,
+    device_config_state = 5002,
+    web_resource_activity = 6001,
+    application_lifecycle = 6002,
+    api_activity = 6003,
+    web_resource_access_activity = 6004
+};
+
+enum class severity_id : int {
+    unknown = 0,
+    informational = 1,
+    low = 2,
+    medium = 3,
+    high = 4,
+    critical = 5,
+    fatal = 6,
+    other = 99,
+};
+
+struct api {
+    ss::sstring operation;
+
+    auto equality_fields() const { return std::tie(operation); }
+};
+
+struct product {
+    ss::sstring name;
+    ss::sstring vendor_name;
+    ss::sstring version;
+
+    auto equality_fields() const {
+        return std::tie(name, vendor_name, version);
+    }
+};
+
+static inline const product& redpanda_product() {
+    static const product redpanda_product_struct{
+      .name = "Redpanda",
+      .vendor_name = ss::sstring{vendor_name},
+      .version = ss::sstring{redpanda_git_version()}};
+    return redpanda_product_struct;
+}
+
+struct metadata {
+    product product;
+    ss::sstring version;
+
+    auto equality_fields() const { return std::tie(product, version); }
+};
+
+static inline const metadata& ocsf_redpanda_metadata() {
+    static const metadata ocsf_metadata{
+      .product = redpanda_product(), .version = ss::sstring{ocsf_api_version}};
+
+    return ocsf_metadata;
+}
+
+struct network_endpoint {
+    fragmented_vector<ss::sstring> intermediate_ips;
+    ss::sstring ip;
+    ss::sstring name;
+    port_t port;
+    ss::sstring svc_name;
+    ss::sstring uid;
+};
+
+struct policy {
+    ss::sstring desc;
+    ss::sstring name;
+
+    auto equality_fields() const { return std::tie(desc, name); }
+};
+
+struct authorization_result {
+    ss::sstring decision;
+    std::optional<policy> policy;
+
+    auto equality_fields() const { return std::tie(decision, policy); }
+};
+
+struct user {
+    enum class type : int {
+        unknown = 0,
+        user = 1,
+        admin = 2,
+        system = 3,
+        other = 99
+    };
+
+    ss::sstring credential_uid;
+    ss::sstring domain;
+    ss::sstring name;
+    type type_id;
+
+    auto equality_fields() const {
+        return std::tie(credential_uid, domain, name, type_id);
+    }
+};
+
+struct actor {
+    fragmented_vector<authorization_result> authorizations;
+    user user;
+
+    auto equality_fields() const { return std::tie(authorizations, user); }
+};
+
+struct resource_detail {
+    ss::sstring name;
+    ss::sstring type;
+
+    auto equality_fields() const { return std::tie(name, type); }
+};
+
+struct authorization_metadata {
+    struct {
+        ss::sstring host;
+        ss::sstring op;
+        ss::sstring permission_type;
+        ss::sstring principal;
+    } acl_authorization;
+
+    struct {
+        ss::sstring name;
+        ss::sstring pattern;
+        ss::sstring type;
+    } resource;
+
+    auto equality_fields() const {
```

Comment:
This is nice, but what about `serde` support? We will need to serialize these structs to iobufs, there is `serde_fields()` for that. I would assume `serde_fields` would look a little different because it would be including things that we didn't want to include in the hash such as the `count` field.

Reply:
We're serializing them to a JSON string not an iobuf right?  The problem is that C++ doesn't have reflection so trying to convert the variable name into a JSON key would require probably some C macro madness.  Maybe I could so something like:

```
auto serde_fields() const {
  return std::tie(std::make_pair("var", _var),....);
}
```

Then have to write a `std::apply` thing.

I already wrote all the JSON serialization code.

Reply:
actually i think i'm wrong here, we just would need to serialize the JSON string straight to an iobuf, we wouldn't be serializing the types themselves.